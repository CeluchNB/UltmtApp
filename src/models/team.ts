import { DisplayUser } from '../types/user'
import { Realm } from '@realm/react'
import { LocalUser, Team } from '../types/team'

export const DisplayTeamSchema: Realm.ObjectSchema = {
    name: 'DisplayTeam',
    embedded: true,
    properties: {
        _id: 'string',
        place: 'string',
        name: 'string',
        teamname: 'string',
        seasonStart: 'string',
        seasonEnd: 'string',
    },
}

export const GuestTeamSchema: Realm.ObjectSchema = {
    name: 'GuestTeam',
    embedded: true,
    properties: {
        _id: 'string?',
        place: 'string?',
        name: 'string',
        teamname: 'string?',
        seasonStart: 'string?',
        seasonEnd: 'string?',
    },
}

export class TeamSchema {
    static schema: Realm.ObjectSchema = {
        name: 'Team',
        primaryKey: '_id',
        properties: {
            _id: 'string',
            place: 'string',
            name: 'string',
            teamname: 'string',
            managers: { type: 'list', objectType: 'DisplayUser' },
            players: { type: 'list', objectType: 'DisplayUser' },
            seasonStart: 'string',
            seasonEnd: 'string',
            seasonNumber: 'int',
            continuationId: 'string',
            rosterOpen: 'bool',
        },
    }

    _id: string
    place: string
    name: string
    teamname: string
    managers: DisplayUser[]
    players: LocalUser[]
    seasonStart: string
    seasonEnd: string
    seasonNumber: number
    continuationId: string
    rosterOpen: boolean

    constructor(team: Team) {
        this._id = team._id
        this.place = team.place
        this.name = team.name
        this.teamname = team.teamname
        this.managers = team.managers
        this.players = team.players
        this.seasonStart = team.seasonStart
        this.seasonEnd = team.seasonEnd
        this.seasonNumber = team.seasonNumber
        this.continuationId = team.continuationId
        this.rosterOpen = team.rosterOpen
    }
}
