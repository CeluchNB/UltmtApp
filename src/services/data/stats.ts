import * as Constants from '../../utils/constants'
import { getInitialPlayerData } from '../../../fixtures/utils'
import { throwApiError } from '../../utils/service-utils'
import {
    AllPlayerStats,
    FilteredGameStats,
    FilteredTeamStats,
    GameStats,
    IdentifiedPlayerStats,
    PlayerStats,
} from '../../types/stats'
import { addPlayerStats, calculatePlayerStats } from '../../utils/stats'
import {
    filterPlayerStats as networkFilterPlayerStats,
    getGameStats as networkGetGameStats,
    getGameStatsByTeam as networkGetGameStatsByTeam,
    getPlayerStats as networkGetPlayerStats,
    getTeamStats as networkGetTeamStats,
    getTeamStatsByGame as networkGetTeamStatsByGame,
} from '../network/stats'

export const getPlayerStats = async (
    id: string,
): Promise<IdentifiedPlayerStats> => {
    try {
        const response = await networkGetPlayerStats(id)
        const { player } = response.data
        return player
    } catch (e) {
        return throwApiError(e, Constants.UNABLE_TO_GET_PLAYER_STATS)
    }
}

export const filterPlayerStats = async (
    id: string,
    teams: string[],
    games: string[],
): Promise<AllPlayerStats> => {
    try {
        const response = await networkFilterPlayerStats(id, teams, games)
        const { stats } = response.data

        const statAggregate: PlayerStats = stats.reduce(
            (a: PlayerStats, b: PlayerStats) => addPlayerStats(a, b),
            getInitialPlayerData({}),
        )
        return calculatePlayerStats(statAggregate)
    } catch (e) {
        return throwApiError(e, Constants.UNABLE_TO_GET_PLAYER_STATS)
    }
}

export const getGameStats = async (id: string): Promise<GameStats> => {
    try {
        const response = await networkGetGameStats(id)
        const { game } = response.data
        return game
    } catch (e) {
        return throwApiError(e, Constants.UNABLE_TO_GET_GAME_STATS)
    }
}

export const getGameStatsByTeam = async (
    gameId: string,
    teamId: string,
): Promise<FilteredGameStats> => {
    try {
        const response = await networkGetGameStatsByTeam(gameId, teamId)
        const { game } = response.data

        return game
    } catch (e) {
        return throwApiError(e, Constants.UNABLE_TO_GET_GAME_STATS)
    }
}

export const getTeamStats = async (
    teamId: string,
): Promise<FilteredTeamStats> => {
    try {
        const response = await networkGetTeamStats(teamId)
        const { team } = response.data

        return team
    } catch (e) {
        return throwApiError(e, Constants.UNABLE_TO_GET_TEAM_STATS)
    }
}

export const getTeamStatsByGame = async (
    teamId: string,
    gameIds: string[],
): Promise<FilteredTeamStats> => {
    try {
        const response = await networkGetTeamStatsByGame(teamId, gameIds)
        const { team } = response.data

        return team
    } catch (e) {
        return throwApiError(e, Constants.UNABLE_TO_GET_TEAM_STATS)
    }
}
