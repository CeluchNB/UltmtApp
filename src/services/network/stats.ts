import { API_KEY, API_URL_V1 } from '@env'
import axios, { AxiosResponse } from 'axios'

export const getPlayerStats = async (id: string): Promise<AxiosResponse> => {
    return await axios.get(`${API_URL_V1}/stats/player/${id}`, {
        headers: {
            'X-API-Key': API_KEY,
        },
    })
}

export const filterPlayerStats = async (
    id: string,
    teams: string[],
    games: string[],
): Promise<AxiosResponse> => {
    let url = `${API_URL_V1}/stats/filter/player/${id}`
    if (teams.length > 0) {
        url += `?teams=${teams.join(',')}`
    }
    if (games.length > 0) {
        if (teams.length > 0) {
            url += `&games=${games.join(',')}`
        } else {
            url += `?games=${games.join(',')}`
        }
    }
    return await axios.get(url, {
        headers: {
            'X-API-Key': API_KEY,
        },
    })
}

export const getGameStats = async (id: string): Promise<AxiosResponse> => {
    return await axios.get(`${API_URL_V1}/stats/game/${id}`, {
        headers: { 'X-API-Key': API_KEY },
    })
}

export const getGameStatsByTeam = async (
    gameId: string,
    teamId: string,
): Promise<AxiosResponse> => {
    return await axios.get(
        `${API_URL_V1}/stats/filter/game/${gameId}?team=${teamId}`,
        { headers: { 'X-API-Key': API_KEY } },
    )
}

export const getTeamStats = async (id: string): Promise<AxiosResponse> => {
    return await axios.get(`${API_URL_V1}/stats/team/${id}`, {
        headers: { 'X-API-Key': API_KEY },
    })
}

export const getTeamStatsByGame = async (
    teamId: string,
    gameIds: string[],
): Promise<AxiosResponse> => {
    let url = `${API_URL_V1}/stats/filter/team/${teamId}`
    if (gameIds.length > 0) {
        url += `?games=${gameIds.join(',')}`
    }
    return await axios.get(url, { headers: { 'X-API-Key': API_KEY } })
}

export const getConnectionStats = async (
    throwerId: string,
    receiverId: string,
): Promise<AxiosResponse> => {
    return await axios.get(
        `${API_URL_V1}/stats/connection?thrower=${throwerId}&receiver=${receiverId}`,
        {
            headers: { 'X-API-Key': API_KEY },
        },
    )
}

export const filterConnectionStats = async (
    throwerId: string,
    receiverId: string,
    games: string[],
    teams: string[],
): Promise<AxiosResponse> => {
    let url = `${API_URL_V1}/stats/filter/connection?thrower=${throwerId}&receiver=${receiverId}`
    if (teams.length > 0) {
        url += `&teams=${teams.join(',')}`
    }
    if (games.length > 0) {
        url += `&games=${games.join(',')}`
    }
    return await axios.get(url, {
        headers: {
            'X-API-Key': API_KEY,
        },
    })
}
