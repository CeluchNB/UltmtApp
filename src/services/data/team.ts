import * as Constants from '../../utils/constants'
import { ApiError } from '../../types/services'
import { throwApiError } from '../../utils/service-utils'
import { CreateTeam, Team } from '../../types/team'
import {
    createTeam as networkCreateTeam,
    getManagedTeam as networkGetManagedTeam,
    getTeam as networkGetTeam,
    removePlayer as networkRemovePlayer,
    rollover as networkRollover,
    searchTeam as networkSearchTeam,
    toggleRosterStatus as networkToggleRosterStatus,
} from '../network/team'

/**
 * Method to create team
 * @param token token of user creating team
 * @param data create team data
 * @returns created team
 * @throws error if team is not created successfully
 */
export const createTeam = async (
    token: string,
    data: CreateTeam,
): Promise<Team> => {
    try {
        const response = await networkCreateTeam(token, data)
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.CREATE_TEAM_ERROR)
    }
}

/**
 * Method to search for teams by place, name, or username
 * @param term search term
 * @returns list of teams matching term
 * @throws error if backend returns an error
 */
export const searchTeam = async (term: string): Promise<Team[]> => {
    try {
        if (term.length < 3) {
            throw new ApiError('Not enough characters to search.')
        }
        const response = await networkSearchTeam(term)
        return response.data
    } catch (error) {
        return throwApiError(error, Constants.SEARCH_ERROR)
    }
}

/**
 * Method to get team that the current user manages
 * @param token jwt of current user
 * @param id team id
 * @returns team object
 * @throws error if backend returns error
 */
export const getManagedTeam = async (
    token: string,
    id: string,
): Promise<Team> => {
    try {
        const response = await networkGetManagedTeam(token, id)
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.GET_TEAM_ERROR)
    }
}

/**
 * Method to get a public team
 * @param id id of team to get
 * @returns team object
 * @throws error if backend returns an error
 */
export const getTeam = async (id: string): Promise<Team> => {
    try {
        const response = await networkGetTeam(id)
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.GET_TEAM_ERROR)
    }
}

/**
 * Method to change a team's roster status
 * @param token jwt of user changing team
 * @param id id of team to update
 * @param open boolean for team's roster status
 * @returns updated team object
 * @throws error if backend returns error
 */
export const toggleRosterStatus = async (
    token: string,
    id: string,
    open: boolean,
): Promise<Team> => {
    try {
        const response = await networkToggleRosterStatus(token, id, open)
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.EDIT_TEAM_ERROR)
    }
}

/**
 * Method to remove a player from a team
 * @param token token of current user
 * @param teamId team player is on
 * @param userId id of player to remove
 * @returns updated team object
 * @throws error if backend returns an error
 */
export const removePlayer = async (
    token: string,
    teamId: string,
    userId: string,
): Promise<Team> => {
    try {
        const response = await networkRemovePlayer(token, teamId, userId)
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.EDIT_TEAM_ERROR)
    }
}

/**
 * Method to start a new season for a team. After this method is called,
 * the old team is no longer editable.
 * @param token jwt of manager of team
 * @param teamId id of team to move over
 * @param copyPlayers boolean for deciding if current roster should be kept
 * @param seasonStart start year for season
 * @param seasonEnd start end for season
 * @returns newly created team object
 * @throws error if backend returns an error
 */
export const rollover = async (
    token: string,
    teamId: string,
    copyPlayers: boolean,
    seasonStart: string,
    seasonEnd: string,
): Promise<Team> => {
    try {
        const response = await networkRollover(
            token,
            teamId,
            copyPlayers,
            seasonStart,
            seasonEnd,
        )
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.EDIT_TEAM_ERROR)
    }
}
