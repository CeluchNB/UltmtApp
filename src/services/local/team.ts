import { Realm } from '@realm/react'
import { Team } from '../../types/team'
import { TeamSchema } from '../../models'
import { getRealm } from '../../models/realm'

export const saveTeams = async (teams: Team[]) => {
    const realm = await getRealm()
    realm.write(() => {
        for (const team of teams) {
            realm.create('Team', team, Realm.UpdateMode.Modified)
        }
    })
}

export const getManagingTeams = async (id: string): Promise<Team[]> => {
    const realm = await getRealm()
    const teams = await realm.objects<TeamSchema>('Team')
    return teams
        .filter(t => t.managers.map(m => m._id).includes(id))
        .map(t => parseTeam(t))
}

const parseTeam = (schema: TeamSchema): Team => {
    return JSON.parse(
        JSON.stringify({
            _id: schema._id,
            place: schema.place,
            name: schema.name,
            teamname: schema.teamname,
            managers: schema.managers,
            players: schema.players,
            seasonStart: schema.seasonStart,
            seasonEnd: schema.seasonEnd,
            seasonNumber: schema.seasonNumber,
            continuationId: schema.continuationId,
            rosterOpen: schema.rosterOpen,
            requests: [],
            games: [],
        }),
    )
}
