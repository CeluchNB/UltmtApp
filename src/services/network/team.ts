import { CreateTeam } from '../../types/team'
import { API_KEY, API_URL_V1 } from '@env'
import axios, { AxiosResponse } from 'axios'

export const createTeam = async (
    token: string,
    data: CreateTeam,
): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/team`,
        {
            team: { ...data },
        },
        {
            headers: {
                Authorization: `Bearer ${token}`,
                'X-API-Key': API_KEY,
            },
        },
    )
}

export const searchTeam = async (
    term: string,
    rosterOpen?: boolean,
): Promise<AxiosResponse> => {
    return await axios.get(
        `${API_URL_V1}/team/search?q=${term}&rosterOpen=${rosterOpen}`,
        {
            headers: { 'X-API-Key': API_KEY },
        },
    )
}

export const getManagedTeam = async (
    token: string,
    id: string,
): Promise<AxiosResponse> => {
    return await axios.get(`${API_URL_V1}/team/managing/${id}`, {
        headers: {
            Authorization: `Bearer ${token}`,
            'X-API-Key': API_KEY,
        },
    })
}

export const getTeam = async (id: string): Promise<AxiosResponse> => {
    return await axios.get(`${API_URL_V1}/team/${id}`, {
        headers: { 'X-API-Key': API_KEY },
    })
}

export const toggleRosterStatus = async (
    token: string,
    id: string,
    open: boolean,
): Promise<AxiosResponse> => {
    return await axios.put(
        `${API_URL_V1}/team/open/${id}?open=${open}`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
                'X-API-Key': API_KEY,
            },
        },
    )
}

export const removePlayer = async (
    token: string,
    teamId: string,
    userId: string,
): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/team/remove/player/${teamId}?user=${userId}`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
                'X-API-Key': API_KEY,
            },
        },
    )
}

export const rollover = async (
    token: string,
    teamId: string,
    copyPlayers: boolean,
    seasonStart: string,
    seasonEnd: string,
): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/team/rollover/${teamId}`,
        {
            copyPlayers,
            seasonStart,
            seasonEnd,
        },
        {
            headers: {
                Authorization: `Bearer ${token}`,
                'X-API-Key': API_KEY,
            },
        },
    )
}

export const addManager = async (
    token: string,
    teamId: string,
    managerId: string,
): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/team/${teamId}/addManager?manager=${managerId}`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
                'X-API-Key': API_KEY,
            },
        },
    )
}

export const getArchivedTeam = async (
    teamId: string,
): Promise<AxiosResponse> => {
    return await axios.get(`${API_URL_V1}/archiveTeam/${teamId}`, {
        headers: {
            'X-API-Key': API_KEY,
        },
    })
}

export const createBulkJoinCode = async (
    token: string,
    teamId: string,
): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/team/getBulkCode?id=${teamId}`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
                'X-API-Key': API_KEY,
            },
        },
    )
}
