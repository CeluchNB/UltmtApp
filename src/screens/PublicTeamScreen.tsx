import * as TeamData from '../services/data/team'
import BaseScreen from '../components/atoms/BaseScreen'
import { Game } from '../types/game'
import { PublicTeamDetailsProps } from '../types/navigation'
import PublicTeamPlayersScene from '../components/organisms/PublicTeamPlayersScene'
import PublicTeamStatsScene from '../components/organisms/PublicTeamStatsScene'
import React from 'react'
import { Team } from '../types/team'
import TeamGameScene from '../components/organisms/TeamGameScene'
import { getGamesByTeam } from '../services/data/game'
import { useTheme } from '../hooks'
import {
    SafeAreaView,
    StyleSheet,
    Text,
    View,
    useWindowDimensions,
} from 'react-native'
import { TabBar, TabView } from 'react-native-tab-view'
import { UseQueryResult, useQuery } from 'react-query'

const renderScene = (
    team: Team,
    error: string,
    loading: boolean,
    teamId: string,
    gamesQuery: UseQueryResult<Game[]>,
    onRefresh: () => Promise<void>,
) => {
    return ({ route }: { route: { key: string } }) => {
        switch (route.key) {
            case 'players':
                return (
                    <PublicTeamPlayersScene
                        team={team}
                        error={error}
                        refreshing={loading}
                        onRefresh={onRefresh}
                    />
                )
            case 'games':
                return (
                    <TeamGameScene teamId={teamId} queryResult={gamesQuery} />
                )
            case 'stats':
                return (
                    <PublicTeamStatsScene
                        teamId={teamId}
                        games={gamesQuery.data || []}
                    />
                )
        }
    }
}

const PublicTeamScreen: React.FC<PublicTeamDetailsProps> = ({
    route,
    navigation,
}) => {
    const { id, archive } = route.params
    const {
        theme: { colors, size, weight },
    } = useTheme()
    const layout = useWindowDimensions()

    const gamesQuery = useQuery(['getGamesByTeam', { teamId: id }], () =>
        getGamesByTeam(id),
    )

    const mapTabNameToIndex = (name: 'players' | 'stats' | 'games'): number => {
        switch (name) {
            case 'players':
                return 0
            case 'games':
                return 1
            case 'stats':
                return 2
        }
    }

    const [team, setTeam] = React.useState({} as Team)
    const [error, setError] = React.useState<string>('')
    const [index, setIndex] = React.useState(mapTabNameToIndex('players'))
    const [loading, setLoading] = React.useState(false)
    const [routes] = React.useState([
        { key: 'players', title: 'Players' },
        { key: 'games', title: 'Games' },
        { key: 'stats', title: 'Stats' },
    ])

    const initializeScreen = React.useCallback(async () => {
        const getTeam = async (): Promise<Team> => {
            if (archive) {
                return TeamData.getArchivedTeam(id)
            }
            return TeamData.getTeam(id)
        }

        setError('')
        setLoading(true)
        getTeam()
            .then(teamResponse => {
                setTeam(teamResponse)
            })
            .catch(e => {
                setError(
                    e.message ??
                        'An error occurred looking for this team. Please try again',
                )
            })
            .finally(() => {
                setLoading(false)
            })
    }, [archive, id])

    React.useEffect(() => {
        const unsubscribe = navigation.addListener('focus', async () => {
            await initializeScreen()
        })

        return unsubscribe
    }, [initializeScreen, navigation])

    React.useEffect(() => {
        navigation.setOptions({ title: `${team.place} ${team.name}` })
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [team])

    const styles = StyleSheet.create({
        screen: {
            height: '100%',
            backgroundColor: colors.primary,
            flex: 1,
        },
        headerContainer: {
            display: 'flex',
            flexDirection: 'column',
            alignSelf: 'center',
            alignItems: 'center',
        },
        date: {
            fontSize: size.fontFifteen,
            fontWeight: weight.medium,
            color: colors.textPrimary,
        },
        teamname: {
            fontSize: size.fontTwenty,
            color: colors.textPrimary,
        },
        bodyContainer: {
            width: '100%',
            alignSelf: 'center',
        },
        error: {
            color: colors.gray,
            fontSize: size.fontThirty,
            fontWeight: weight.bold,
            textAlign: 'center',
        },
    })

    return (
        <BaseScreen containerWidth={95}>
            <View style={styles.headerContainer}>
                <Text style={styles.teamname}>@{team?.teamname}</Text>
                {team?.seasonStart === team?.seasonEnd ? (
                    <Text style={styles.date}>
                        {new Date(team?.seasonStart || '').getUTCFullYear()}
                    </Text>
                ) : (
                    <Text style={styles.date}>
                        {new Date(team?.seasonStart || '').getUTCFullYear() +
                            ' - ' +
                            new Date(team?.seasonEnd || '').getUTCFullYear()}
                    </Text>
                )}
            </View>
            <SafeAreaView
                style={{
                    // TODO: check this height on multiple devices
                    height: layout.height - 225,
                }}>
                <TabView
                    navigationState={{ index, routes }}
                    renderScene={renderScene(
                        team,
                        error,
                        loading,
                        id,
                        gamesQuery,
                        initializeScreen,
                    )}
                    swipeEnabled={false}
                    onIndexChange={setIndex}
                    initialLayout={{ width: layout.width }}
                    renderTabBar={props => {
                        return (
                            <TabBar
                                {...props}
                                style={{ backgroundColor: colors.primary }}
                                indicatorStyle={{
                                    backgroundColor: colors.textPrimary,
                                }}
                                activeColor={colors.textPrimary}
                                inactiveColor={colors.darkGray}
                            />
                        )
                    }}
                />
            </SafeAreaView>
        </BaseScreen>
    )
}

export default PublicTeamScreen
