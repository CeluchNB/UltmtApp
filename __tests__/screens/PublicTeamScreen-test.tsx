import * as StatsData from '../../src/services/data/stats'
import * as TeamData from '../../src/services/data/team'
import { NavigationContainer } from '@react-navigation/native'
import { PublicTeamDetailsProps } from '../../src/types/navigation'
import PublicTeamScreen from '../../src/screens/PublicTeamScreen'
import React from 'react'
import { Team } from '../../src/types/team'
import { GameStats, TeamStats } from '../../src/types/stats'
import { QueryClient, QueryClientProvider } from 'react-query'
import { act, fireEvent, render, screen } from '@testing-library/react-native'

jest.mock('react-native-element-dropdown', () => {
    return {
        Dropdown: () => <span>dropdown</span>,
    }
})

const client = new QueryClient()

const navigate = jest.fn()
const addListener = jest.fn()
const setOptions = jest.fn()
const mockedNavigate = jest.fn()
jest.mock('@react-navigation/native', () => {
    const actualNav = jest.requireActual('@react-navigation/native')
    return {
        ...actualNav,
        useNavigation: () => ({
            navigate: mockedNavigate,
        }),
    }
})
jest.mock('react-native-gifted-charts', () => {
    return {
        BarChart: () => {},
        __esModule: true,
    }
})

const props: PublicTeamDetailsProps = {
    navigation: {
        navigate,
        addListener,
        setOptions,
        isFocused: () => true,
    } as any,
    route: {
        params: { id: 'id1', place: 'Place', name: 'Name' },
    } as any,
}

const playerOne = {
    _id: 'user1',
    firstName: 'First 1',
    lastName: 'Last 1',
    username: 'firstlast1',
}
const gameStats: GameStats = {
    _id: 'game1',
    startTime: '01/01/2023',
    teamOneId: 'team1',
    points: [],
    goalsLeader: {
        player: playerOne,
        total: 1,
    },
    assistsLeader: {
        player: playerOne,
        total: 1,
    },
    blocksLeader: {
        player: playerOne,
        total: 1,
    },
    turnoversLeader: {
        player: playerOne,
        total: 1,
    },
    plusMinusLeader: {
        player: playerOne,
        total: 1,
    },
    pointsPlayedLeader: {
        player: playerOne,
        total: 1,
    },
    momentumData: [],
}

const team: Team = {
    _id: 'id1',
    place: 'Place',
    name: 'Name',
    teamname: 'placename',
    managers: [],
    players: [
        {
            _id: 'id',
            firstName: 'First',
            lastName: 'Last',
            username: 'firstlast',
        },
    ],
    seasonNumber: 1,
    seasonStart: '2022',
    seasonEnd: '2022',
    continuationId: 'id123',
    rosterOpen: true,
    requests: [],
}

const teamStats: TeamStats = {
    ...team,
    players: [],
    games: [],
    winPercentage: 1,
    offensiveConversion: 0.8,
    defensiveConversion: 0.4,
    wins: 10,
    losses: 0,
    goalsFor: 100,
    goalsAgainst: 67,
    holds: 87,
    breaks: 13,
    turnoverFreeHolds: 45,
    offensePoints: 23,
    defensePoints: 54,
    turnovers: 4,
    turnoversForced: 45,
    completionsToScore: [],
    completionsToTurnover: [],
}

beforeEach(() => {
    navigate.mockReset()
    addListener.mockReset()
})

beforeAll(() => {
    jest.useFakeTimers({ legacyFakeTimers: true })
})
afterAll(() => {
    jest.useRealTimers()
})

it('should handle player click', async () => {
    const spy = jest
        .spyOn(TeamData, 'getTeam')
        .mockReturnValueOnce(Promise.resolve(team))
    jest.spyOn(StatsData, 'getTeamStats').mockImplementation(() => {
        return Promise.resolve({
            ...teamStats,
            ...team,
            ...gameStats,
            players: [],
        })
    })

    render(
        <NavigationContainer>
            <QueryClientProvider client={client}>
                <PublicTeamScreen {...props} />
            </QueryClientProvider>
        </NavigationContainer>,
    )
    await act(async () => {
        const scrollView = screen.getByTestId('public-team-scroll-view')
        const { refreshControl } = scrollView.props
        refreshControl.props.onRefresh()
    })

    const playerView = screen.getByText(
        `${team.players[0].firstName} ${team.players[0].lastName}`,
    )
    expect(screen.getAllByText('Players').length).toBeGreaterThan(1)
    expect(screen.getAllByText('Stats').length).toBeGreaterThan(1)
    expect(screen.getAllByText('Years').length).toBeGreaterThan(0)

    fireEvent.press(playerView)
    expect(mockedNavigate).toHaveBeenCalled()
    spy.mockRestore()
})

it('should handle get team error', async () => {
    const spy = jest
        .spyOn(TeamData, 'getTeam')
        .mockReturnValueOnce(Promise.reject({ message: 'error' }))
    jest.spyOn(StatsData, 'getTeamStats').mockReturnValue(
        Promise.resolve({
            ...teamStats,
            ...team,
            ...gameStats,
            players: [],
        }),
    )
    render(
        <NavigationContainer>
            <QueryClientProvider client={client}>
                <PublicTeamScreen {...props} />
            </QueryClientProvider>
        </NavigationContainer>,
    )

    await act(async () => {
        const scrollView = screen.getByTestId('public-team-scroll-view')
        const { refreshControl } = scrollView.props
        refreshControl.props.onRefresh()
    })

    const errorText = screen.getByText('error')
    expect(screen.getAllByText('Players').length).toBeGreaterThan(1)
    expect(screen.getAllByText('Stats').length).toBeGreaterThan(1)

    expect(errorText).toBeTruthy()
    spy.mockRestore()
})

it('should get archive team and handle player click', async () => {
    const spy = jest
        .spyOn(TeamData, 'getArchivedTeam')
        .mockReturnValueOnce(Promise.resolve(team))
    jest.spyOn(StatsData, 'getTeamStats').mockReturnValue(
        Promise.resolve({
            ...teamStats,
            ...team,
            ...gameStats,
            players: [],
        }),
    )
    const tempProps: PublicTeamDetailsProps = {
        navigation: {
            navigate,
            addListener,
            setOptions,
            isFocused: () => true,
        } as any,
        route: {
            params: { id: 'id1', place: 'Place', name: 'Name', archive: true },
        } as any,
    }
    render(
        <NavigationContainer>
            <QueryClientProvider client={client}>
                <PublicTeamScreen {...tempProps} />
            </QueryClientProvider>
        </NavigationContainer>,
    )

    await act(async () => {
        const scrollView = screen.getByTestId('public-team-scroll-view')
        const { refreshControl } = scrollView.props
        refreshControl.props.onRefresh()
    })

    const playerView = screen.getByText(
        `${team.players[0].firstName} ${team.players[0].lastName}`,
    )

    expect(screen.getAllByText('Players').length).toBeGreaterThan(1)
    expect(screen.getAllByText('Stats').length).toBeGreaterThan(1)

    fireEvent.press(playerView)
    expect(mockedNavigate).toHaveBeenCalled()
    spy.mockRestore()
})
