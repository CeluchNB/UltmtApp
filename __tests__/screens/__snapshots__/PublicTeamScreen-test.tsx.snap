// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
      "height": "100%",
    }
  }
>
  <RCTScrollView
    refreshControl={
      <RefreshControlMock
        colors={
          Array [
            "#ffd600",
          ]
        }
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    testID="public-team-scroll-view"
  >
    <RCTRefreshControl />
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#3183ff",
                "fontSize": 82.1917808219178,
                "fontWeight": "bold",
              },
              Object {
                "textAlign": "center",
              },
            ]
          }
        >
          Place Name
        </Text>
        <Text
          style={
            Object {
              "color": "#3183ff",
              "fontSize": 30.821917808219176,
              "fontWeight": "600",
              "marginBottom": 5,
              "textAlign": "center",
            }
          }
        >
          NaN
        </Text>
        <Text
          style={
            Object {
              "color": "#3183ff",
              "fontSize": 30.821917808219176,
              "marginBottom": 5,
              "textAlign": "center",
            }
          }
        >
          @
        </Text>
      </View>
      <View
        style={
          Object {
            "alignSelf": "center",
            "width": "75%",
          }
        }
      >
        <View
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "display": "flex",
                "flex": 1,
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#3183ff",
                  "flex": 1,
                  "fontSize": 61.64383561643835,
                  "fontWeight": "bold",
                  "marginBottom": 10,
                }
              }
            >
              Players
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;

exports[`should match snapshot with get archive team 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
      "height": "100%",
    }
  }
>
  <RCTScrollView
    refreshControl={
      <RefreshControlMock
        colors={
          Array [
            "#ffd600",
          ]
        }
        onRefresh={[Function]}
        refreshing={false}
      />
    }
    testID="public-team-scroll-view"
  >
    <RCTRefreshControl />
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#3183ff",
                "fontSize": 82.1917808219178,
                "fontWeight": "bold",
              },
              Object {
                "textAlign": "center",
              },
            ]
          }
        >
          Place Name
        </Text>
        <Text
          style={
            Object {
              "color": "#3183ff",
              "fontSize": 30.821917808219176,
              "fontWeight": "600",
              "marginBottom": 5,
              "textAlign": "center",
            }
          }
        >
          NaN
        </Text>
        <Text
          style={
            Object {
              "color": "#3183ff",
              "fontSize": 30.821917808219176,
              "marginBottom": 5,
              "textAlign": "center",
            }
          }
        >
          @
        </Text>
      </View>
      <View
        style={
          Object {
            "alignSelf": "center",
            "width": "75%",
          }
        }
      >
        <View
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <View
            style={
              Object {
                "display": "flex",
                "flex": 1,
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#3183ff",
                  "flex": 1,
                  "fontSize": 61.64383561643835,
                  "fontWeight": "bold",
                  "marginBottom": 10,
                }
              }
            >
              Players
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
